File: Typescript/2024/2024-010/2024-010.ts
import * as fs from 'fs';

const INPUT_FILE = '\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-010\\input.txt';

function readMap(filename: string): number[][] {
    const lines = fs.readFileSync(filename, 'utf-8').split('\n').filter(line => line.trim() !== '');
    return lines.map(line => line.trim().split('').map(ch => parseInt(ch, 10)));
}

function neighbors(r: number, c: number, rows: number, cols: number): [number, number][] {
    const directions: [number, number][] = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    return directions
        .map(([dr, dc]) => [r + dr, c + dc] as [number, number])
        .filter(([nr, nc]) => nr >= 0 && nr < rows && nc >= 0 && nc < cols);
}

function findTrailheadScores(grid: number[][]): number {
    const rows = grid.length;
    const cols = grid[0].length;

    const trailheads: [number, number][] = [];
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (grid[r][c] === 0) {
                trailheads.push([r, c]);
            }
        }
    }

    let totalScore = 0;

    for (const [startR, startC] of trailheads) {
        const visited = new Set<string>();
        const queue: [number, number][] = [];
        const reachableNines = new Set<string>();

        queue.push([startR, startC]);
        visited.add(`${startR},${startC}`);

        while (queue.length > 0) {
            const [r, c] = queue.shift()!;
            const currentHeight = grid[r][c];

            if (currentHeight === 9) {
                reachableNines.add(`${r},${c}`);
            } else {
                const nextHeight = currentHeight + 1;
                for (const [nr, nc] of neighbors(r, c, rows, cols)) {
                    if (!visited.has(`${nr},${nc}`) && grid[nr][nc] === nextHeight) {
                        visited.add(`${nr},${nc}`);
                        queue.push([nr, nc]);
                    }
                }
            }
        }

        totalScore += reachableNines.size;
    }

    return totalScore;
}

function countPaths(
    r: number,
    c: number,
    grid: number[][],
    dp: (number | null)[][],
    rows: number,
    cols: number
): number {
    if (dp[r][c] !== null) {
        return dp[r][c]!;
    }

    const currentHeight = grid[r][c];

    if (currentHeight === 9) {
        dp[r][c] = 1;
        return 1;
    }

    let totalPaths = 0;
    const nextHeight = currentHeight + 1;
    for (const [nr, nc] of neighbors(r, c, rows, cols)) {
        if (grid[nr][nc] === nextHeight) {
            totalPaths += countPaths(nr, nc, grid, dp, rows, cols);
        }
    }

    dp[r][c] = totalPaths;
    return totalPaths;
}

function calculateTotalRating(grid: number[][]): number {
    const rows = grid.length;
    const cols = grid[0].length;

    const trailheads: [number, number][] = [];
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (grid[r][c] === 0) {
                trailheads.push([r, c]);
            }
        }
    }

    const dp: (number | null)[][] = Array.from({ length: rows }, () => Array(cols).fill(null));

    return trailheads.reduce((total, [tr, tc]) => total + countPaths(tr, tc, grid, dp, rows, cols), 0);
}

function main() {
    const startTimePart1 = performance.now();
    const grid = readMap(INPUT_FILE);
    const totalScore = findTrailheadScores(grid);
    const endTimePart1 = performance.now();

    console.log(`Part 1 Result: ${totalScore}`);
    console.log(`Time taken for Part 1: ${(endTimePart1 - startTimePart1) / 1000}s`);

    const startTimePart2 = performance.now();
    const totalRating = calculateTotalRating(grid);
    const endTimePart2 = performance.now();

    console.log(`Part 2 Result: ${totalRating}`);
    console.log(`Time taken for Part 2: ${(endTimePart2 - startTimePart2) / 1000}s`);
}

main();

File: Ruby/2024/2024-010/2024-010.rb
INPUT_FILE = '\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-010\\input.txt'

def read_map(filename)
  grid = []
  File.readlines(filename).each do |line|
    line.strip!
    next if line.empty?
    grid << line.chars.map(&:to_i)
  end
  grid
end

def neighbors(r, c, rows, cols)
  directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
  directions.each_with_object([]) do |(dr, dc), result|
    nr, nc = r + dr, c + dc
    result << [nr, nc] if nr.between?(0, rows - 1) && nc.between?(0, cols - 1)
  end
end

def find_trailhead_scores(grid)
  rows = grid.size
  cols = rows > 0 ? grid[0].size : 0

  trailheads = []
  rows.times do |r|
    cols.times do |c|
      trailheads << [r, c] if grid[r][c] == 0
    end
  end

  total_score = 0

  trailheads.each do |start_r, start_c|
    visited = {}
    queue = [[start_r, start_c]]
    visited[[start_r, start_c]] = true

    reachable_nines = {}

    until queue.empty?
      r, c = queue.shift
      current_height = grid[r][c]

      if current_height == 9
        reachable_nines[[r, c]] = true
      else
        next_height = current_height + 1
        neighbors(r, c, rows, cols).each do |nr, nc|
          next if visited[[nr, nc]] || grid[nr][nc] != next_height
          visited[[nr, nc]] = true
          queue << [nr, nc]
        end
      end
    end

    total_score += reachable_nines.keys.size
  end

  total_score
end

def count_paths(r, c, grid, dp, rows, cols)
  return dp[r][c] unless dp[r][c].nil?

  current_height = grid[r][c]

  if current_height == 9
    dp[r][c] = 1
    return 1
  end

  total_paths = 0
  next_height = current_height + 1
  neighbors(r, c, rows, cols).each do |nr, nc|
    if grid[nr][nc] == next_height
      total_paths += count_paths(nr, nc, grid, dp, rows, cols)
    end
  end

  dp[r][c] = total_paths
  total_paths
end

def calculate_total_rating(grid)
  rows = grid.size
  cols = rows > 0 ? grid[0].size : 0

  trailheads = []
  rows.times do |r|
    cols.times do |c|
      trailheads << [r, c] if grid[r][c] == 0
    end
  end

  dp = Array.new(rows) { Array.new(cols) }

  total_rating = 0
  trailheads.each do |tr, tc|
    total_rating += count_paths(tr, tc, grid, dp, rows, cols)
  end

  total_rating
end

def main
  start_time_part1 = Time.now
  grid = read_map(INPUT_FILE)
  total_score = find_trailhead_scores(grid)
  end_time_part1 = Time.now
  puts "Part 1 Result: #{total_score}"
  puts "Time taken for Part 1: #{(end_time_part1 - start_time_part1).round(9)} s"

  start_time_part2 = Time.now
  total_rating = calculate_total_rating(grid)
  end_time_part2 = Time.now
  puts "Part 2 Result: #{total_rating}"
  puts "Time taken for Part 2: #{(end_time_part2 - start_time_part2).round(9)} s"
end

main
